### ISSUE TITLE
Add a free-form GDN normalization layer (FFF-style, learned inverse, approximate likelihood)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/gauss_flows#116]
> **Type:** enhancement (new module + training mode)
> **Area:** new `src/gauss_flows/_src/free_form/` package + extension to `src/gauss_flows/_src/inference/train.py`
> **Status:** proposal — ready for implementation
> **Sibling issue:** #115 — *exact-likelihood* `GeneralizedDivisiveNormalization` (same transform, dense log-det). **This issue depends on #115 landing first** so the encoder can reuse the same parametric block.
> **Overlap check:** searched the 41 open issues for `gdn | divisive | free.?form | fff | hutchinson | injective | rademacher`; no overlap.

---

## Summary

Add **Generalized Divisive Normalization (GDN)** as a **free-form-flow (FFF) layer**: use the same GDN block from #115 as an expressive *encoder* `f_θ`, learn a separate **IGDN-style decoder** `g_φ` as its approximate inverse, and train the pair under a maximum-likelihood objective whose log-det term is a **Hutchinson trace surrogate** — decoder Jacobian standing in for the inverse encoder Jacobian — with invertibility enforced by a **reconstruction loss** rather than by computing GDN's expensive dense Jacobian determinant. This follows Draxler et al., *Free-form Flows* (AISTATS 2024) / Sorrenson et al. (ICLR 2024).

The point: GDN's exact log-det is a dense `D × D` determinant (O(D³)) — fine in channel mode (see #115), prohibitive for large full coupling. The FFF route sidesteps it entirely, giving an **O(D) training step** and **single-evaluation sampling**, at the cost of **only approximate likelihood and only empirical invertibility**. This is explicitly the "free-form flows" box rather than the classical exact-Gaussianisation box — worth having as a contrast/option alongside the exact bijector.

This is **not** an `AbstractBijection` (which would promise exact log-det) — it is a `flowjax.distributions.AbstractDistribution`-shaped wrapper around an `encoder + decoder` pair, with its own training loop.

---

## Motivation & user stories

* **As a user who wants GDN's divisive-normalisation expressivity at scale** (large coupling, high-D image tensors), I want to train it without the O(D³) determinant, accepting an approximate likelihood.
* **As a user generating samples**, I want **one-step** sampling via the learned IGDN decoder (no fixed-point iteration at sample time).
* **As a researcher**, I want a drop-in comparison point to the exact GDN bijector (#115) and to the package's RBIG / coupling flows — same transform, different likelihood/invertibility trade-off.
* **As a contributor**, I want it to reuse the existing `matfree.stochtrace` Hutchinson plumbing (already in `_src/_divergence.py`), the existing `fit_gaussianization_flow` skeleton for training (already in `_src/inference/train.py`), standard equinox modules for the decoder, and `jaxtyping` annotations — passing the four CI gates.

---

## Background: how FFF applies to GDN

FFF trains an encoder `f_θ` (here a GDN block from #115) and decoder `g_φ` (here an IGDN-style net) with the per-event loss

```
L(x)  =  ½ ‖f_θ(x)‖²                    (Gaussian-latent NLL)
      −  vᵀ J_θ(x) · stop_grad(J_φ(f_θ(x)) v)      (log-det surrogate)
      +  β · ½ ‖x − g_φ(f_θ(x))‖²        (reconstruction)
```

* The **first** term is the standard normal base NLL on the latent.
* The **second** is the log-det surrogate: `tr(J_φ J_θ)` estimated by one Hutchinson probe `v`, with the decoder JVP `J_φ v` standing in for `J_θ⁻¹ v`. One forward + one backward pass — **O(D)**, no determinant.
* The **third** is the reconstruction loss enforcing `g_φ ≈ f_θ⁻¹`; at the optimum the surrogate is exact (error bounded by `‖J_θ J_φ − I‖_F`).

Sampling: `x = g_φ(z)`, `z ∼ 𝒩(0, I)` — one decoder call. Exact per-point likelihood (if needed at eval) still requires the decoder Jacobian determinant, computed only at evaluation time.

Contrast with #115: there, invertibility and likelihood are **guaranteed / exact** but the determinant is O(D³) and the inverse is iterative; here, both are **approximate** but training is O(D) and sampling is one step.

---

## Existing project conventions to match

| Topic | Precedent in repo | Reference |
|---|---|---|
| Hutchinson trace surrogate | `matfree.stochtrace` with `_ProbeDistribution = Literal["rademacher", "normal"]` and a `_make_sampler` helper. Use the same idiom — don't roll our own Rademacher draw. | `src/gauss_flows/_src/_divergence.py:42, 88` |
| Training loop | `fit_gaussianization_flow` wraps `flowjax.train.loops.fit_to_data`. For FFF we cannot use the default NLL loss directly — we need a custom `loss_fn` parameter or a hand-rolled `optax` loop. | `src/gauss_flows/_src/inference/train.py:10` |
| Distribution wrapping | `flowjax.distributions.AbstractDistribution`; `gaussianization_flow` returns one. To plug into the package's info-theoretic utilities (`entropy`, `mutual_information`, etc.) the FFF wrapper should be one too. | `src/gauss_flows/_src/flows.py` |
| GDN block | Will be added by #115 (`GeneralizedDivisiveNormalization`, `GeneralizedDivisiveNormalization1D`). The encoder here is literally that block. | (depends on #115) |
| `eqx.Module` style | `_src/transforms/bijections/normalization/actnorm.py` shows the convention: shape field, `ClassVar[None]` cond_shape, equinox-typed array fields, `__init__(shape, …)`. | `src/gauss_flows/_src/transforms/bijections/normalization/actnorm.py` |
| Test layout | `tests/test_train.py` for training utilities; one `tests/test_*.py` per new module otherwise. | `tests/test_train.py` |
| Re-export chain | `_src/free_form/__init__.py` → `src/gauss_flows/__init__.py`; the training utility is re-exported through the existing `inference` chain. | `src/gauss_flows/__init__.py` |
| `einx` | Not a current project dep. Don't introduce. `jnp` / `jnp.einsum` for contractions. | `pyproject.toml` |

---

## API proposal

New file: `src/gauss_flows/_src/free_form/gdn.py`. Companion training helper extends `src/gauss_flows/_src/inference/train.py`.

```python
from typing import ClassVar, Literal
from flowjax.distributions import AbstractDistribution
from jaxtyping import Array, ArrayLike, Float
import equinox as eqx


class FreeFormGDN(AbstractDistribution):
    """Free-form Gaussianization with a GDN encoder + IGDN-style decoder.

    Trained with the FFF (Draxler et al. 2024) loss: latent-Gaussian NLL +
    Hutchinson log-det surrogate + reconstruction. Approximate per-point
    likelihood; one-step sampling.

    Args:
        encoder: GDN block from #115 (channel- or 1-D coupling).
        decoder: Inverse-normalisation network (parametric IGDN by default,
            or a free-form MLP / 1x1-conv).
        decoder_kind: One of {"igdn", "mlp", "conv1x1"}; controls how the
            companion factory builds the default decoder.
    """
    shape: tuple[int, ...]
    cond_shape: ClassVar[None] = None
    encoder: eqx.Module          # GeneralizedDivisiveNormalization[1D] from #115
    decoder: eqx.Module
    decoder_kind: Literal["igdn", "mlp", "conv1x1"] = "igdn"

    # AbstractDistribution surface
    def _log_prob(self, x, condition=None) -> Float[Array, ""]: ...   # eval-time exact via decoder Jacobian
    def _sample(self, key, condition=None) -> Float[Array, "..."]:    # one decoder pass
        z = jr.normal(key, self.shape)
        return self.decoder(z)


def free_form_gdn_loss(
    model: FreeFormGDN,
    x: Float[Array, "..."],
    key,
    *,
    beta: float = 100.0,
    probe: Literal["rademacher", "normal"] = "rademacher",
) -> Float[Array, ""]:
    """Per-event FFF loss; vmap over batch."""


def fit_free_form_gdn(
    key,
    model: FreeFormGDN,
    data: ArrayLike,
    *,
    beta: float = 100.0,
    learning_rate: float = 5e-4,
    max_epochs: int = 100,
    max_patience: int = 5,
    batch_size: int = 256,
    val_prop: float = 0.1,
    show_progress: bool = True,
) -> tuple[FreeFormGDN, dict[str, list[float]]]:
    """FFF training loop — same signature shape as fit_gaussianization_flow.

    Uses a custom loss; cannot delegate to flowjax.train.fit_to_data directly.
    A hand-rolled optax loop (mirroring the structure of fit_to_data) is the
    cleanest fit.
    """
```

Public re-export chain (3 files):

```
src/gauss_flows/_src/free_form/__init__.py      # NEW
src/gauss_flows/_src/inference/__init__.py      # re-export fit_free_form_gdn alongside fit_gaussianization_flow
src/gauss_flows/__init__.py                     # public API
```

---

## Implementation sketch — surrogate via `matfree.stochtrace`

Mirror `hutchinson_divergence` (`_divergence.py:42`) for the probe sampler, but route the probe through one **encoder JVP** and one **decoder JVP** instead of one vector-field JVP:

```python
import jax
import jax.numpy as jnp
from matfree import stochtrace


def _free_form_loss_per_event(model, x, key, beta, probe):
    # Sampler matches hutchinson_divergence's convention.
    if probe == "rademacher":
        sampler = stochtrace.sampler_rademacher(jnp.zeros_like(x), num=1)
    else:
        sampler = stochtrace.sampler_normal(jnp.zeros_like(x), num=1)
    v = sampler.sample(key)[0]                                  # (D,)

    # Two JVPs and a reconstruction; no determinant anywhere.
    z, jv_enc = jax.jvp(model.encoder, (x,), (v,))              # f(x), J_θ v
    _, jv_dec = jax.jvp(model.decoder, (z,), (jv_enc,))         # J_φ J_θ v
    surrogate = jnp.sum(v * jax.lax.stop_gradient(jv_dec))      # vᵀ J_φ J_θ v
    nll_latent = 0.5 * jnp.sum(z * z)
    x_rec = model.decoder(z)
    recon = 0.5 * jnp.sum((x - x_rec) ** 2)
    return nll_latent - surrogate + beta * recon


def free_form_gdn_loss(model, x_batch, key, beta=100.0, probe="rademacher"):
    keys = jax.random.split(key, x_batch.shape[0])
    return jax.vmap(_free_form_loss_per_event, in_axes=(None, 0, 0, None, None))(
        model, x_batch, keys, beta, probe
    ).mean()
```

**Why not delegate to `flowjax.train.loops.fit_to_data`?** That helper hard-codes the NLL loss; the FFF objective adds the surrogate and reconstruction terms, so a hand-rolled `optax` loop (mirroring the structure of `fit_to_data`) is the cleanest fit. The signature of `fit_free_form_gdn` should match `fit_gaussianization_flow` so users have one familiar API.

**`β` schedule.** The FFF paper recommends a warm-up on `β`: start low so the encoder learns useful structure before reconstruction dominates, then raise so reconstruction is small enough to make the likelihood trustworthy and to exclude the degenerate `f = g = 0` solution. Default `β = 100`, linear warm-up over `~10%` of epochs, exposed as a kwarg.

---

## Tests

Add to `tests/test_train.py` (existing file) for the training side; new file `tests/test_free_form.py` for the model side.

```python
# tests/test_free_form.py

def test_free_form_gdn_shape(key):
    """encode / decode preserve the event shape."""

def test_free_form_gdn_one_step_sampling(key):
    """sample is a single decoder call (no while_loop), jit/vmap-safe."""

def test_free_form_gdn_surrogate_finite(key):
    """Hutchinson surrogate is finite and gradient flows to both encoder and
    decoder params (i.e. stop_gradient on decoder JVP doesn't kill the encoder
    grad)."""

def test_free_form_gdn_recon_drives_invertibility(key):
    """Train a few steps on a small problem; reconstruction loss decreases
    and ‖J_θ J_φ − I‖_F shrinks accordingly."""

def test_free_form_gdn_surrogate_matches_exact_on_small_D(key):
    """On small D where #115's exact GDN log-det is cheap, the FFF surrogate
    gradient w.r.t. encoder params aligns (cosine ≈ 1) with the exact log-det
    gradient once reconstruction is near zero."""


# tests/test_train.py — extend

def test_fit_free_form_gdn_smoke(key):
    """fit_free_form_gdn runs end-to-end on a tiny synthetic dataset,
    returns (model, losses) with both 'train' and 'val' keys; loss decreases."""
```

---

## TODOs until completion

* [ ] **Land #115 first.** The encoder here is the `GeneralizedDivisiveNormalization` block from #115; building this without it would duplicate the forward and softplus-reparam logic.
* [ ] New package `src/gauss_flows/_src/free_form/` with `gdn.py` containing `FreeFormGDN` (subclasses `flowjax.distributions.AbstractDistribution` so it plugs into `entropy`, `mutual_information`, etc. for evaluation).
* [ ] Decoder default: **parametric IGDN** (`y_i · (β'_i + Σ_j γ'_ij y_j²)^{1/2}`) mirroring GDN's own fixed-point inverse — low param count, structural inductive bias matching the encoder. Also implement `decoder_kind="mlp"` and `decoder_kind="conv1x1"` factories for fully free-form decoders.
* [ ] `free_form_gdn_loss`: `matfree.stochtrace` Rademacher / normal probe (mirror `_divergence.py:88`), `jax.jvp` for encoder and decoder, `jax.lax.stop_gradient` on the decoder JVP only.
* [ ] Eval-time `log_prob` via decoder Jacobian determinant (O(D³) at eval — document clearly).
* [ ] `fit_free_form_gdn` in `_src/inference/train.py` with `fit_gaussianization_flow`'s signature shape; hand-rolled `optax` loop because the loss is not pure NLL.
* [ ] `β` warm-up schedule: linear over the first 10% of epochs by default, exposed as `beta_warmup_epochs`.
* [ ] Diagnostics: track reconstruction error and `‖J_θ J_φ − I‖_F` per epoch; warn if eval-time `log_prob` is invoked while reconstruction is above a configurable threshold.
* [ ] Tests as above + `jit`/`vmap`/`grad` smoke on CPU; float32 and float64.
* [ ] Image path: GDN/IGDN as `1×1` channel coupling inside an equinox `Conv` wrapper; small CelebA/CIFAR-patch smoke test in `docs/notebooks/` (next free index — current head is `10_information_theory.ipynb`).
* [ ] Pass `pytest`, `ruff check .`, `ruff format --check .`, `ty check src/gauss_flows`. Update `CHANGELOG.md`. Add an entry to the API docs linking to #115 as the exact-likelihood sibling.

---

## Acceptance criteria

* Training reduces reconstruction below a set tolerance on a synthetic problem; one-step sampling works under `jit`/`vmap`.
* On small D (where #115's exact log-det is cheap), FFF surrogate gradients align with exact GDN gradients (cosine ≥ 0.95 once reconstruction is below tol) — confirms the surrogate is doing what FFF says it does.
* `FreeFormGDN` plugs into `entropy`, `mutual_information`, etc. as an `AbstractDistribution` (evaluation only; sampling is exact, `log_prob` is approximate).
* Documentation explicitly states likelihood is approximate and invertibility is empirical, with a pointer to #115 for the exact-likelihood / exact-invertibility variant.
* All four pre-commit gates pass.

---

## What datasets does this make sense for?

* **High-D / large-coupling image tensors:** the main reason to choose FFF over #115's exact path — channel- or spatially-coupled GDN at a scale where the O(D³) determinant is prohibitive, with one-step sampling. Natural fit for image patches, learned-compression latents, and image-feature density.
* **When sampling speed matters:** one decoder pass vs the exact layer's fixed-point inverse.
* **Not ideal when you need certified exact likelihood or guaranteed invertibility** (information-theoretic estimation, exact density on tabular / scientific data) — use #115 or the RBIG / marginal blocks. FFF likelihoods are approximate and only trustworthy once reconstruction is driven low.

---

## References

* Draxler, Sorrenson, Zimmermann, Rousselot & Köthe. *Free-form Flows: Make Any Architecture a Normalizing Flow.* AISTATS 2024. arXiv:2310.16624.
* Sorrenson, Draxler, Rousselot, Hummerich, Zimmermann & Köthe. *Lifting Architectural Constraints of Injective Flows.* ICLR 2024. arXiv:2306.01843.
* Keller, Peters, Jaini, Hoogeboom, Forré & Welling. *Self-Normalizing Flows.* ICML 2021. arXiv:2011.07248.
* Ballé, Laparra & Simoncelli. *Density Modeling of Images using a Generalized Normalization Transformation.* ICLR 2016. arXiv:1511.06281; and *End-to-end Optimized Image Compression*, ICLR 2017 (IGDN).
* `matfree` — used here for the Hutchinson trace surrogate, matching the existing pattern in `src/gauss_flows/_src/_divergence.py::hutchinson_divergence`.
* Sibling issue #115 — exact-likelihood `GeneralizedDivisiveNormalization`.

---

## What changed vs. earlier drafts of this proposal

* **Area path corrected** from `_src/bijections/` (does not exist) to a new package `src/gauss_flows/_src/free_form/` plus extension of `src/gauss_flows/_src/inference/train.py`. The class is *not* an `AbstractBijection` (no exact log-det), so it doesn't belong under `bijections/`.
* **Hutchinson estimator** routed through `matfree.stochtrace` (already a project dep, already used in `_src/_divergence.py:42–101`) instead of a hand-rolled Rademacher draw.
* **Training utility** wired to mirror `fit_gaussianization_flow`'s signature but with a hand-rolled `optax` loop rather than `flowjax.train.fit_to_data` (which is NLL-only).
* **`FreeFormGDN`** typed as a `flowjax.distributions.AbstractDistribution` (not a plain `eqx.Module`) so it plugs straight into `entropy` / `mutual_information` and friends.
* **`einx`** demoted from required to "not adding a dep for it" — same call as in #115.
* **Encoder reuse** of #115's block flagged as a hard sequencing dependency; this issue should only be started after #115 ships.
* **Notebook index** for the smoke-test notebook pinned at the actual next free slot (current head: `10_information_theory.ipynb`).
